### PR TITLE
pacific: tool/bluestore-tool: introduce repair_omap_upgrade command (continuation)

### PR DESCRIPTION
**Pull request by aclamk**
This is the improvement / continuation of original repair command:
https://github.com/ceph/ceph/pull/43820

Major features:
- restrictive data check before key modification
- state machine to detect omap head copies
- state machine to detect object discontinuity (just report, not fixing anything)
- counts problems during repair
- --dry-run option

Missing:
- actual procedure to test in repeatable way, including a way to run broken fsck-quick-fix ?

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests
- Teuthology
  - [ ] Completed teuthology run
  - [ ] No teuthology test necessary (e.g., documentation)

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -8287,56 +8287,189 @@ int BlueStore::repair_omap_upgrade()
...
-        if (cnt >= 1024) {
-          key = it->key();
-          db->submit_transaction_sync(txn);
-          it->lower_bound(key);
-          txn = db->get_transaction();
-          cnt = 0;
-        }
+	dout(20) << __func__ << " key=" << pretty_binary_string(key) << dendl;
+	it->next();
+	ceph_assert(key.length() >= 8 + 4 + 8 + 1);
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:8309]: let's just report invalid key and continue here....

**Comment by ifed01** [comment]
@aclamk - haven't read through the patch in details yet. Just a couple comments so far:
1) This looks like a bit of overkill. My plan was to have a simple and quick fix to help in those few cases we've got. And then throw this code away... 
Given that a sort of combination of my original patch and your initial proposal to match a few fixed patterns seems sufficient.

2) If we go this more complex way it would be great to have an ability to run simple (i.e. other than traversing the real DB) UTs for this state machine/matching algorithm. So may be detach that state machine logic into a standalone class which gets the input from the formal interface? I've just made the first step in my original - detached the key matching func and wrote some UTs to verify it. What's about using the similar approach here?

File: src/os/bluestore/BlueStore.cc (hunk 1 above)
Comment (aclamk) [reply src/os/bluestore/BlueStore.cc:8309]: right....

**Comment by aclamk** [comment]
@ifed01 I agree with your sentiment: "overkill".
It is certainly much more code then I expected at the beginning.
I would love to make it as short as possible, but without sacrificing confidence that conversion is not injecting additional errors.

**Comment by aclamk** [comment]
Replaced by: #43917

**Pull request closed without merging**